### PR TITLE
Skeleton for the assignment-6

### PR DESCRIPTION
**Pull request by ndomanova**
Basic implementation of the rearrangeCars() function.
Next things to do:
-add consistency check
-improve readability of moveCar() function
-add javadocs
-add tests

Implementation of the rearrangeCarsLeastMoves() function(optional challange 1).
To do:
-combine two functions moveCar(...)

Basic implementation of ReservedParking class(optional challenge 3), which needs renaming of the methods, tests, javadocs and improve behavior in the corner cases(e.g. when there is no lot to put the car).

### CODE REVIEW

File: group/assignment-6/Parking.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        addRecord(moves, car, carPosition, emptySpacePosition);
+    }
+
+    private void moveCar(Map<Character, Integer> state, Map<Integer, Character> inverseState, char car, StringBuilder moves) {
+        int emptySpacePosition = state.get(emptyLot);
+        int carPosition = state.get(car);
+
+        state.put(car, emptySpacePosition);
+        state.put(emptyLot, carPosition);
+        inverseState.put(emptySpacePosition, car);
```
Comment (Utena) [inline group/assignment-6/Parking.java]: Repeated code? This function is very similar as the previous one. Maybe, we can merge both and use a boolean variable when the `inverseState` map has to be employed.

File: group/assignment-6/Parking.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    }
+
+    public String rearrangeCarsLeastMoves(Map<Character, Integer> endState) {
+        int counter = 0;
+        StringBuilder moves = new StringBuilder();
+        Map<Character, Integer> currentState = copyMap(initialState);
+        Map<Integer, Character> inverseEndState = getInverseMap(endState);
+
+        while (counter < endState.size()-1) {
```
Comment (Utena) [inline group/assignment-6/Parking.java]: Follow the Google Java Style Guide: `while (counter < endState.size() - 1) {`, so add a space between - 1.

File: group/assignment-6/Parking.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+
+    public String rearrangeCars(Map<Character, Integer> endState) {
+
+        // add checking for consistency initial and end states of parking!
+        // e.g. if there is no empty lot in the state
+        // or the size of end state is not equal to the size of the initial state
+
+        Map<Integer, Character> inverseCurrentState = getInverseMap(initialState);
```
Comment (Utena) [inline group/assignment-6/Parking.java]: I think we need to explain in the Javadoc why you are inverting the `initialState`.

File: group/assignment-6/Parking.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                int currentPosition = currentState.get(currentCar);
+                int endPosition = lot.getValue();
+
+                if (currentPosition != endPosition) {
+                    char carAtEndPosition = getCarAtPosition(inverseCurrentState, endPosition);
+
+                    if (carAtEndPosition == emptyLot) {
+                        moveCar(currentState, inverseCurrentState, currentCar, moves);
+                    } else {
+                        moveCar(currentState, inverseCurrentState,  carAtEndPosition, moves);
```
Comment (Utena) [inline group/assignment-6/Parking.java]: To improve readability and if you want to preserve two `moveCar` functions, it will be a good choice to implement both `moveCarToEmptyLot` and `moveCar`. What do you think?

File: group/assignment-6/Main.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        endState.put('C', 5);
+        endState.put('D', 1);
+        endState.put('E', 4);
+        endState.put('F', 3);
+
+        String moves = parking.rearrangeCars(endState);
+        String efficientMoves = parking.rearrangeCarsLeastMoves(endState);
+
+        System.out.println(moves);
+        System.out.println();
```
[the remark this answers is not in the document]
Comment (ndomanova) [reply group/assignment-6/Main.java]: Sure! :)

File: group/assignment-6/Parking.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    private Map<Character, Integer> initialState;
+
+    public Parking(Map<Character, Integer> state) {
+        this.initialState = state;
+    }
+
+
+    public String rearrangeCars(Map<Character, Integer> endState) {
+
+        // add checking for consistency initial and end states of parking!
```
[the remark this answers is not in the document]
Comment (ndomanova) [reply group/assignment-6/Parking.java]: Totally agree with you, that's why I put all of those comments :)

File: group/assignment-6/Parking.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    private char getCarAtPosition(Map<Integer, Character> state, int position) {
+        return state.get(position);
+    }
+
+
+    private Map<Integer, Character> getInverseMap(Map<Character, Integer> map) {
+
+        Map<Integer, Character> inverseMap = new HashMap<Integer, Character>();
```
[the remark this answers is not in the document]
Comment (ndomanova) [reply group/assignment-6/Parking.java]: I was thinking about using stream, but couldn't arrange it(didn't know about Collectors.toMap()). Thanks a lot!

File: group/assignment-6/Parking.java (hunk 1 above)
Comment (ndomanova) [reply group/assignment-6/Parking.java]: Yup, I wrote about it in the description of the pull request, so we need to merge it somehow.

File: group/assignment-6/Parking.java (hunk 4 above)
Comment (ndomanova) [reply group/assignment-6/Parking.java]: I'm not pretty sure, because every time when we invoke `moveCar` we actually move it to the empty lot, so it can be confusing.
But we can discuss it next meeting :)

File: group/assignment-6/Parking.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+        state.put(car, emptySpacePosition);
+        state.put(emptyLot, carPosition);
+        inverseState.put(emptySpacePosition, car);
+        inverseState.put(carPosition, emptyLot);
+
+        addRecord(moves, car, carPosition, emptySpacePosition);
+    }
+
+    private void addRecord(StringBuilder moves, char car, int source, int destination) {
```
[the remark this answers is not in the document]
Comment (ndomanova) [reply group/assignment-6/Parking.java]: But when you concatenate Strings like this, you increase your space complexity. That's the only reason why I use StringBuilder.

File: group/assignment-6/Parking.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Parking {
```
[the remark this answers is not in the document]
Comment (ndomanova) [reply group/assignment-6/Parking.java]: That's a perfect idea! Done!

File: group/assignment-6/Parking.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+        for (Map.Entry<T, S> lot: finalState.entrySet()) {
+            T car = lot.getKey();
+
+            if (car != emptySpace) {
+                S currentLot = currentState.get(car);
+                S desiredLot = lot.getValue();
+
+                if (currentLot != desiredLot) {
+                    T carAtDesiredLot = getCarAtLot(inverseCurrentState, desiredLot);
```
Comment (angelina-nedelcu) [inline group/assignment-6/Parking.java]: We can remove this function. It is quite readable that one line that replaces it.

File: group/assignment-6/Parking.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+                    if (carAtDesiredLot == emptySpace) {
```
Comment (angelina-nedelcu) [inline group/assignment-6/Parking.java]: Both branches have the same line of code. This means that we can change the way it is written this way:

if (carAtDesiredlot != emptySpace) {
    moveCar(currentstate, inversecurrentstate, caratdesiredlot, moves);
}
//and here put the duplicated line
movecar(currentState, inversecurrentstate, car, moves);

File: group/assignment-6/Parking.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                    return moves.toString();
+                }
+            }
+            moveCar(currentState, desiredCar, moves);
+
+        }
+        return moves.toString();
+    }
+
+    private void moveCar(Map<T, S> state,
```
Comment (angelina-nedelcu) [inline group/assignment-6/Parking.java]: Nice function and implementation!

File: group/assignment-6/Parking.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+        return new HashMap<>(parkingState);
+    }
+
+    protected T getCarAtLot(Map<S, T> state, S lot) {
+        return state.get(lot);
+    }
+
+    protected Map<S, T> getInverseMap(Map<T, S> map) {
+
+        return map.entrySet().stream().collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));
```
Comment (angelina-nedelcu) [inline group/assignment-6/Parking.java]: Very nice line!

File: group/assignment-6/Parking.java [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                }
+
+            }
+        }
+
+        return moves.toString();
+
+    }
+
+    public String rearrangeCarsLeastMoves(Map<T, S> endState) {
```
Comment (angelina-nedelcu) [inline group/assignment-6/Parking.java]: Great idea!

File: group/assignment-6/ReservedParking.java [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        Map<S, T> inverseFinalState = getInverseMap(finalState);
+
+        while (carsAtCorrectLot < finalState.size() - 1) {
+            S emptySpaceLot = currentState.get(emptySpace);
+            T desiredCar = inverseFinalState.get(emptySpaceLot);
+
+            if (desiredCar != emptySpace && lotIsReservedForCar(emptySpaceLot, desiredCar)) {
+                carsAtCorrectLot++;
+            } else {
+
```
Comment (angelina-nedelcu) [inline group/assignment-6/ReservedParking.java]: I think we should remove the space here.

File: group/assignment-6/ReservedParking.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        int carsAtCorrectLot = 0;
+        StringBuilder moves = new StringBuilder();
+        Map<T, S> currentState = copyMap(initialState);
+        Map<S, T> inverseFinalState = getInverseMap(finalState);
+
+        while (carsAtCorrectLot < finalState.size() - 1) {
+            S emptySpaceLot = currentState.get(emptySpace);
+            T desiredCar = inverseFinalState.get(emptySpaceLot);
+
+            if (desiredCar != emptySpace && lotIsReservedForCar(emptySpaceLot, desiredCar)) {
```
Comment (angelina-nedelcu) [inline group/assignment-6/ReservedParking.java]: This means that is a possibilty that the final state does not respect the constraints?

File: group/assignment-6/Parking.java (hunk 13 above)
Comment (ndomanova) [reply group/assignment-6/Parking.java]: Thanks to @Utena :)

File: group/assignment-6/Parking.java (hunk 11 above)
Comment (ndomanova) [reply group/assignment-6/Parking.java]: Thanks! I didn't notice it!

File: group/assignment-6/ReservedParking.java (hunk 16 above)
Comment (ndomanova) [reply group/assignment-6/ReservedParking.java]: Actually, I just realized `lotIsReservedForCar()` is useless in this statement. So, basically, we can check if the final state does not respect the constraints here or at the very beginning.
